### ISSUE TITLE
Make the compositor render software cursors

### ISSUE THREAD

**Issue by emersion** [swaywm/wlroots#1363]
Making wlroots responsible for rendering them causes some issues. It adds rendering code when swapping buffers, after `wlr_renderer_end`. It causes issues like https://github.com/swaywm/wlroots/issues/1291, because `wlr_output` is not presentation-time-aware (and shouldn't be).

We should probably remove the rendering code from `wlr_output_swap_buffers` and make the compositor do it. We could provide a helper function to render software cursors that could be called by the compositor in its rendering process.

Same applies to fullscreen surfaces -- but we should remove `wlr_output_set_fullscreen_surface` altogether, replacing it with something else.

@ascent12 How does this interact with the renderer redesign?

---

wlroots has migrated to gitlab.freedesktop.org. This issue has been moved to:

https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/1363

**Comment by ascent12** [comment]
+1
Rendering it for the user was a naive decision I made before we had things like damage handling and redrew the entire frame every time. It doesn't interact nicely now that we actually have those extra features.

>How does this interact with the renderer redesign?
It's one of those things I need to look into. My knowledge of how X11 actually handles cursors isn't great, so I'll have to see what's actually possible which would work over every backend.

Ideally we'd create another wlr_image that you could draw to with normal renderer commands.

**Comment by emersion** [comment]
>My knowledge of how X11 actually handles cursors isn't great, so I'll have to see what's actually possible which would work over every backend.

On X11 we only support software cursors. I don't think it's worth it to do anything more complicated.

**Comment by ascent12** [comment]
True. Since with wayland and DRM, we can basically just use normal buffers (although allocated with a special parameter for DRM), it should be able to fit pretty easily into wlr_allocator.
I'll see if X11 can take an xcb_pixmap_t and directly use it as a cursor or not. If it can't we'll just keep ignoring it.

**Comment by ammen99** [comment]
I think this is a good idea. Having the compositor draw software cursors has another advantage as well - wlroots won't be adding damage (AFAIK) so that the compositor is free to track damage in any way it wants to. Currently wlroots forces me to use scaled coordinates for damage which I find, well, quite inconvenient and complicates damage handling in wayfire.

**Comment by emersion** [comment]
>Currently wlroots forces me to use scaled coordinates for damage which I find, well, quite inconvenient and complicates damage handling in wayfire.

After this change wlroots would take damage in renderer coordinates (scaled + transformed), so that it can provide it directly to the backend.

**Comment by ammen99** [comment]
> > Currently wlroots forces me to use scaled coordinates for damage which I find, well, quite inconvenient and complicates damage handling in wayfire.
> 
> After this change wlroots would take damage in renderer coordinates (scaled + transformed), so that it can provide it directly to the backend.

You mean when swapping buffers? If yes, that's not a problem, I need to implement transforming from layout to scaled+transformed coordinates anyway for the final phase of drawing, but coordinate systems will be consistent throughout the rest of the code.

**Comment by emersion** [comment]
Remaining issues to get this closed:

- [ ] wlroots sends frame_done events to cursor surfaces in `wlr_output_swap_buffers`. This makes it impossible for compositors to implement repaint scheduling for cursor surfaces.
- [ ] wlroots still damages software cursors when moved (see `output_cursor_damage_whole`, we probably want to remove `wlr_output_damage_whole` too)

**Comment by ascent12** [comment]
I think that we need to remove handling of software cursors **entirely** from wlroots, including any damage tracking for them.
Basically, we aren't the ones in control of the `wlr_renderer` and the scene graph, so us performing any rendering really ends up causing issues and requires `wlr_output` to be a very complex interface for giving damage information etc. back to the user.
Compositors should draw the cursors themselves with the renderer functions directly, and opportunistically make use of hardware cursors.

- [ ] Remove `struct wlr_output_cursor` and all related functions.
- [ ] Remove any functions that cause rendering from `struct wlr_cursor`. I still think it's a useful interface for tracking the cursor position on `struct wlr_output_layout`, but should not actually render anything.
- [ ] Remove `wlr_xcursor_manager_set_cursor_image`.
- [ ] Work out how "cursor locks" are going to work with the screen recording interfaces. Probably just involves the compositor listening to the signals themselves.
- [ ] Maybe some other stuff I'm forgetting

The `wlr_output_cursor_set_*` function needs to be rethought, which is dependent on how we'd add support for planes. Maybe that could be left until later, and just revert to being a function that works directly on the `wlr_output`.

**Comment by emersion** [comment]
+1 overall. We might want to keep some helpers for cursors though, but definitely not in `wlr_output`.

Additionally it's not clear how to handle cursors in the Wayland backend, especially with multi-seat.

>revert to being a function that works directly on the wlr_output

That could be a good step to push things forwards, yes.

**Comment by ddevault** [comment]
I agree that we should offer a helper, outside of `wlr_output`.
